expat/lib/xmlparse.c

```diff
@@ -5438,6 +5486,26 @@ processInternalEntity(XML_Parser parser, ENTITY *entity,
       parser->m_freeInternalEntities = openEntity;
     }
   }
+
+  if (!(parser->m_options & XML_OPTION_HUGE_ENTITES)) {
+    XML_Index index = XML_GetCurrentByteIndex(parser);
+    parser->m_entityRecursionLevel--;
+    if (entity->textLen > XML_ENTITY_EXPANSION_SIZE) {
+      /* Current entity is too large */
+      return XML_ERROR_ENTITY_EXPANSION_LIMIT;
+    }
+    /* Entity expension length isn't equal to output length. For nested
+     * entities, it records length of each expansion level and not sum of
+     * expanded text. */
+    parser->m_entityExpansionLen += entity->textLen;
+    if ((index > 0) &&
+        (parser->m_entityExpansionLen > XML_ENTITY_EXPANSION_SIZE) &&
+        (parser->m_entityExpansionLen > index * XML_ENTITY_EXPANSION_RATIO)) {
+      /* Ratio between processed XML bytes and all expanded entities is off */
+      return XML_ERROR_ENTITY_EXPANSION_LIMIT;
```

Comment:
Code `XML_ERROR_ENTITY_EXPANSION_LIMIT` seems to be used for two (slightly) different error scenarios now. Is there some security concern why we need to obscure those two cases into one? Else I imagine debugging is going to be more fun if those two cases can be told apart. What do you think?